tests/test_clean.py

```diff
@@ -1,96 +1,210 @@
-    TAGS = ['span', 'br']
-    ATTRS = {'span': ['style']}
+@pytest.mark.parametrize('data, should_strip, expected', [
+    # Regular comment
+    (
+        '<!-- this is a comment -->',
+        True,
+        ''
+    ),
 
-    text = 'a <br/><span style="color:red">test</span>'
-    assert (
-        bleach.clean(text, tags=TAGS, attributes=ATTRS) ==
-        'a <br><span style="">test</span>'
+    # Open comment with no close comment bit
+    (
+        '<!-- open comment',
+        True,
+        ''
+    ),
+    (
+        '<!--open comment',
+        True,
+        ''
+    ),
+    (
+        '<!-- open comment',
+        False,
+        '<!-- open comment-->'
+    ),
+    (
+        '<!--open comment',
+        False,
+        '<!--open comment-->'
+    ),
+
+    # Comment with text to the right
+    (
+        '<!-- comment -->text',
+        True,
+        'text'
+    ),
+    (
+        '<!--comment-->text',
+        True,
+        'text'
+    ),
+    (
+        '<!-- comment -->text',
+        False,
+        '<!-- comment -->text'
+    ),
+    (
+        '<!--comment-->text',
+        False,
+        '<!--comment-->text'
+    ),
+
+    # Comment with text to the left
+    (
+        'text<!-- comment -->',
+        True,
+        'text'
+    ),
+    (
+        'text<!--comment-->',
+        True,
+        'text'
+    ),
+    (
+        'text<!-- comment -->',
+        False,
+        'text<!-- comment -->'
+    ),
+    (
+        'text<!--comment-->',
+        False,
+        'text<!--comment-->'
     )
+])
+def test_comments(data, should_strip, expected):
+    assert clean(data, strip_comments=should_strip) == expected
 
 
-def test_named_arguments():
-    ATTRS = {'a': ['rel', 'href']}
+@pytest.mark.parametrize('data, expected', [
+    # Disallowed tag is escaped
+    ('<img src="javascript:alert(\'XSS\');">', '&lt;img src="javascript:alert(\'XSS\');"&gt;'),
+
+    # Test with parens
+    ('a <script>safe()</script> test', 'a &lt;script&gt;safe()&lt;/script&gt; test'),
+
+    # Test with braces
+    ('a <style>body{}</style> test', 'a &lt;style&gt;body{}&lt;/style&gt; test'),
+])
+def test_disallowed_tags(data, expected):
+    assert clean(data) == expected
 
-    text = '<a href="http://xx.com" rel="alternate">xx.com</a>'
-    assert bleach.clean(text) == '<a href="http://xx.com">xx.com</a>'
+
+def test_invalid_char_in_tag():
+    # NOTE(willkg): Two possible outcomes because attrs aren't ordered
+    assert (
+        clean('<script/xss src="http://xx.com/xss.js"></script>') in
+        [
+            '&lt;script src="http://xx.com/xss.js" xss=""&gt;&lt;/script&gt;',
+            '&lt;script xss="" src="http://xx.com/xss.js"&gt;&lt;/script&gt;'
+        ]
+    )
     assert (
-        bleach.clean(text, attributes=ATTRS) ==
-        '<a href="http://xx.com" rel="alternate">xx.com</a>'
+        clean('<script/src="http://xx.com/xss.js"></script>') ==
+        '&lt;script src="http://xx.com/xss.js"&gt;&lt;/script&gt;'
```

Comment:
I inlined the `text` argument. I find it's easier to read and reason about when the in and out are next to each other.

Reply:
👍